org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/SnippetCompletionProposal.java

```diff
@@ -571,13 +572,30 @@ private static CompletionItem getRecordSnippet(SnippetCompletionContext scc, IPr
 
 	private static void setFields(CompletionItem ci, ICompilationUnit cu, CompletionItemDefaults completionItemDefaults) {
 		ci.setKind(CompletionItemKind.Snippet);
-		ci.setInsertTextFormat(InsertTextFormat.Snippet);
-		if (completionItemDefaults.getInsertTextFormat() != null && completionItemDefaults.getInsertTextFormat() == InsertTextFormat.Snippet){
-			ci.setInsertTextFormat(null);
-		}
+		setInsertTextFormat(ci, completionItemDefaults);
+		setInsertTextMode(ci, completionItemDefaults);
 		ci.setDocumentation(SnippetUtils.beautifyDocument(ci.getInsertText()));
 	}
 
+	private static void setInsertTextFormat(final CompletionItem item, CompletionItemDefaults completionItemDefaults) {
+		if (!JavaLanguageServerPlugin.getPreferencesManager().getClientPreferences().isCompletionListItemDefaultsInsertTextFormatSupport() ||
+			completionItemDefaults.getInsertTextFormat() == null ||
+			completionItemDefaults.getInsertTextFormat() != InsertTextFormat.Snippet
+		) {
+			item.setInsertTextFormat(InsertTextFormat.Snippet);
+		}
+	}
+
+	private static void setInsertTextMode(final CompletionItem item, CompletionItemDefaults completionItemDefaults) {
```

Comment:
Such kind of utilities (setInsertTextFormat, setInsertTextMode, etc...) can be reused from different kind of completion handlers. Maybe we can extract it as a utility.